File: 1339-maximum-product-of-splitted-binary-tree/1339-maximum-product-of-splitted-binary-tree.java
class Solution {
    long total = 0;
    long maxP = 0;
    
    public int sum(TreeNode root) {
        if (root == null) return 0;

        int sumLeft = sum(root.left);
        int sumRight = sum(root.right);
        long subtreeSum = root.val + sumLeft + sumRight;
        
        long remSubtreeSum = total - subtreeSum;
        maxP = Math.max(maxP, subtreeSum * remSubtreeSum);
        
        return (int)subtreeSum;
    }


    public int maxProduct(TreeNode root) {

        if (root == null) return 0;

        total = sum(root);
        maxP = 0;
        
        sum(root);
        return (int)((maxP) % (long)(1e9+7));
    }
}

File: 1339-maximum-product-of-splitted-binary-tree/1339-maximum-product-of-splitted-binary-tree.cpp
class Solution {
public:
    int calculateTotalSum(TreeNode* root){
        
        if(root == NULL) return 0;
        
        int left = calculateTotalSum(root->left);
        int right = calculateTotalSum(root->right);
        
        return root->val + left + right;
    }
    
    int findMaximumProduct(TreeNode* root, long totalSum, long& maxP){
        
        if(root == NULL) return 0;
        
        int left = findMaximumProduct(root->left, totalSum, maxP);
        int right = findMaximumProduct(root->right, totalSum, maxP);
        long subtreeSum = root->val + left + right;
        
        long remainingSubtreeSum = totalSum - subtreeSum;
        maxP = max(maxP, subtreeSum * remainingSubtreeSum);
        
        return subtreeSum;
    }
    
    int maxProduct(TreeNode* root) {
        
        long totalSum = calculateTotalSum(root);
        long maxP = 0;
        
        findMaximumProduct(root, totalSum, maxP);
        
        int MOD = 1e9+7;
        return maxP % MOD;
    }
};